crates/red_knot_server/src/session.rs

```diff
@@ -262,7 +264,12 @@ impl DocumentSnapshot {
     }
 
     pub(crate) fn file(&self, db: &RootDatabase) -> Option<File> {
-        let path = url_to_system_path(self.document_ref.file_url()).ok()?;
-        system_path_to_file(db, path).ok()
+        match url_to_any_system_path(self.document_ref.file_url()).ok()? {
+            AnySystemPath::System(path) => system_path_to_file(db, path).ok(),
+            AnySystemPath::SystemVirtual(virtual_path) => db
+                .files()
+                .try_virtual_file(&virtual_path)
+                .map(|virtual_file| virtual_file.file()),
+        }
     }
```

Comment:
I thought about adding `File` to the `DocumentController` but my concern is about notebook cells which also uses `TextDocument` to represent them. I'm not exactly sure how to represent them. I do want to follow-up on this when adding support for notebooks.

Reply:
Ah okay :)